sw/device/tests/uart_parity_test.c

```diff
@@ -0,0 +1,233 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_clkmgr.h"
+#include "sw/device/lib/dif/dif_lc_ctrl.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/clkmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_console.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/ottf_utils.h"
+#include "sw/device/lib/testing/test_framework/status.h"
+#include "sw/device/lib/testing/uart_testutils.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_pinmux_t pinmux;
+static dif_rv_plic_t rv_plic;
+static dif_uart_t uart;
+
+// Parameters of the test to be set at the beginning.
+static volatile uint8_t parity = 0xff;
+static volatile uint8_t uart_idx = 0xff;
+
+// There are multiple uart instances in the chip. These variables will be
+// updated according to the uart we select.
+static volatile uint32_t uart_peripheral_id;
+static volatile uint32_t uart_irq_rx_parity_err_id;
+
+// Whether we expect and have received the RX parity error interrupt.
+static volatile bool uart_irq_rx_parity_err_expected = false;
+static volatile bool uart_irq_rx_parity_err_fired = false;
+
+enum {
+  kCommandTimeout = 5000000,  // microseconds
+};
+
+typedef enum test_phase {
+  kTestPhaseCfg,
+  kTestPhaseSend,
+  kTestPhaseRecv,
+  kTestPhaseRecvErr,
+  kTestPhaseDone,
+} test_phase_t;
+static volatile uint8_t test_phase = kTestPhaseCfg;
+
+// Some random bytes that will be sent and received to check the parity.
+static const uint8_t kUartData[32] = {
+    0x3f, 0x39, 0xb0, 0x4e, 0xa6, 0xce, 0xe5, 0xb7, 0x94, 0x48, 0xec,
+    0xb5, 0x48, 0x5c, 0x08, 0x5b, 0xcd, 0x47, 0xae, 0x80, 0xbb, 0x49,
+    0xa1, 0x7c, 0x39, 0x20, 0xd1, 0x6d, 0x2f, 0x4f, 0x94, 0xd8,
+};
+
+// Configure the UART under test selected by `uart_idx` with `parity`.
+static status_t configure_uart(void) {
```

Comment:
There is a similar (but more complex) function in the uart tx/rx test, maybe this could be moved to a test util?

Reply:
I did think about this since it's copied from that test, however I wasn't a fan of having the test_util share its global state (the statics containing the interrupts etc) with tests. Maybe all of the interrupts and base addresses could be put into some structs which the testutil returns, but that was also a little complex.

Do you mind if I leave this duplicated in the test for now? I'm writing another test for line breaks which might need something similar so I will consider extracting it for that test

Reply:
Sure, I was thinking a struct but maybe it's a bit overkill.

Reply:
On second thought, I've added it to this PR instead.